### PR TITLE
Issue #82 solution

### PR DESCRIPTION
**Pull request by Dinh-Hung-Tu**
replace regex match by regex search for detection of invalid characters in the middle of the string.

### CODE REVIEW

**Comment by Dinh-Hung-Tu** [comment]
@kirsle Can you please take a review on this?

File: rivescript/parser.py [hunk 1]
```diff
@@ -535,13 +535,13 @@ def check_syntax(self, cmd, line):
...
+                search = re.search(RE.name_syntax, line)
+                if search:
                     return "Topics should be lowercased and contain only numbers and letters"
             elif parts[0] == "object":
-                match = re.match(RE.name_syntax, line)
-                if match:
-                    return "Objects can only contain numbers and letters"
+                search = re.search(RE.name_syntax, line)
+                if search:
+                    return "Objects can only contain numbers and letters"  # TODO Acceptance of uppercase letter?
```
Comment (kirsle) [inline rivescript/parser.py]: It seems safe enough to me to allow uppercase letters in object names. You may need to add a new syntax regexp to `regexp.py` to support that, like `r'[^A-Za-z0-9_\-\s]'`

File: tests/test_format_message.py [hunk 2]
```diff
@@ -28,4 +28,26 @@ def test_format_triggers(self):
...
+                                + hello
+                                - hi
+                            <topics
+                            """)
+        self.assertRaises(Exception, self.new, """
+                                    > object hash %perl
+                                      my ($rs, $args) = @_;
+                                      my $method = shift @{$args};
+                                    <object
+                                    """)  # Test for character violation in object, no %
```
Comment (kirsle) [inline tests/test_format_message.py]: I don't like how all these blocks of code are formatted. The indentations are all over the place.

Maybe something like this instead:

```python
    def test_invalid_character_raise_exception(self):
        # This test passes with `match`, which only check at the beginning
        self.assertRaises(Exception, self.new, """
            + $hello
            - hi
        """)

        # etc for all the other tests.
```

**Comment by kirsle** [comment]
Just to be sure: this change makes the `$` character invalid for triggers right? I'd want it to raise a RiveScript syntax error at parse time.

`$` is a special metacharacter in regexps and I don't have a custom use for it in RiveScript like I do for some of the other metacharacters (e.g. `[` and `]`), so I'd want it to raise an error instead of silently doing the wrong thing when the regexp gets executed.

File: rivescript/parser.py (hunk 1 above)
Comment (Dinh-Hung-Tu) [inline rivescript/parser.py]: I add this accordingly.

**Comment by Dinh-Hung-Tu** [comment]
Yes, $ is one of many invalid characters. The valid sets is  ( | ) [ ] * _ # @ { } < > = and it is check at https://github.com/aichaos/rivescript-python/blob/master/rivescript/parser.py#L594. I may made a mistake putting a review there. I don't know how to quote the code to here. You can close that if not an appropriate place.

**Comment by Dinh-Hung-Tu** [comment]
@kirsle Do you plan to include this fix in the new release anytime soon? I find this bug fix is urgent because `check_syntax` checks the validity of the trigger. For example, `$1.00` has already been detected as incorrect format but `1.00$` slipped through the check because the string doesn't start with invalid character.

**Comment by kirsle** [comment]
@Dinh-Hung-Tu I just published v1.14.6 to PyPI so you can update to get this fix. :smile:

**Comment by Dinh-Hung-Tu** [comment]
Thanks @kirsle!

**Pull request merged**